include/oneapi/mkl/bfloat16.hpp

```diff
@@ -0,0 +1,228 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#ifndef _BFLOAT16_HPP__
+#define _BFLOAT16_HPP__
+
+#include <cmath>
+#include <cstdint>
+#include <type_traits>
+
+namespace oneapi {
+namespace mkl {
+
+namespace bfloat16_impl {
```

Comment:
As far as I know there is no C++ language standard for bf16, so users can have their own custom type to work with bf16 feature. With introducing public bf16 type here users will have to cast their bf16 types to this one to work with oneMKL and back to work with custom kernels. Did you consider some void* approach to simplify this work for users?

Reply:
That is a good suggestion. My concern is, I am not sure how to test it properly without a sample custom type. Maybe do this change when/if we need it?

Reply:
Can we add this custom type to our tests only and refer to them as an expected type to use for new API?

Reply:
Since Intel oneMKL uses this type, I guess we don't have other choices but define the same type in opensource oneMKL.